### ISSUE TITLE
Performance Updates

### ISSUE THREAD

**Issue by JasonBristol** [benbaptist/minecraft-wrapper#295]
I have been working on cleaning up our threads for a few days now and have encountered some pain points that I will be resolving before I feel comfortable pushing any kind of threading updates. 

These issues do not necessarily block any threading work, but will make it easier to rework our threads. This set of work involves performance improvements, refactoring/rewriting, and stability improvements. Ultimately these work toward giving us a more robust code set.

**None** of these changes involve fundamental changes to our current code set in regards to features or capabilities.

The following needs to be addressed:

- **Exception Handling**
 - To better understand and capture any threading issues we need a more robust approach to error handling. This includes, but is not limited to providing either a stack trace or an error message tuple. Many of our errors are reported via an abstracted message that provides no context into the issue. These messages are logged and are typically only viewable by server administrators so we shouldn't be shy.
- **File Handles**
  - All of our file handles are left open for the garbage collector. Although python is pretty good at cleaning this up, it is bad practice to leave our handles open, especially with our high volume of reads and writes. I would hate to track how many handles we could have open in a 10 minute span of running the server, but I would not be surprised if it was in the triple digits. We need to manage these file handles responsibly, this may no doubt help out with #276 
- **Performance Improvements & Code Cleanup**
  - Much of our code set is iterative and drawn out. This actually makes it very difficult to read and understand. Any two given areas appear to be written in completely different conventions etc. I have identified a few areas where we can make improvements that may not provide a noticeable performance adjustment, but will provide more readable code and standardize our approach across our code base.
- **Deprecated Libraries**
  - I understand the intention of reducing external dependencies, but many of our libraries are quite frankly deprecated or superseded by a better library. Why are we using os.walk and not scandir.walk? Why aren't we using the logging package pre bundled in python? Ultimately I would like to replace and/or strip out any deprecated libraries in favor of more modern libraries. We can address the cost-benefit of external dependencies when we encounter them.

There may be more, but for right now these are the big things sticking out at me. I will be moving my work stream over to these tasks in preparation for reworking our threads. Each of these will be benefited by spinning off separate issues for each and assigning them to me so I can track them appropriately. This may require @suresttexas00 help since he is a collaborator.

If there are any items I may have missed or any concerns please feel free to track them in this issue thread.

I will begin this work once #292 is merged (@benbaptist any updates on this?) so that I have the most current code set before going through everything.

**Comment by suresttexas00** [comment]
#292 is ready for pulling

**Comment by JasonBristol** [comment]
Unless there are concerns, I will also be standardizing all of Wrapper.py to UTF-8 encoding. I know this was mentioned in #43 & #250 briefly, but I think it is time to settle. This will also help toward a python 3 port as python 3 is much more explicit than python 2

**Comment by suresttexas00** [comment]
Both of those issues (#43 and 250) were about the the "encoding" of the wrapper.py file.. which is not a true *.py file, but is actually an archive.  Python 2 understands this and unpacks it, but Python 3 uses no such construct; so python 3 cannot open wrapper.py... As far as conversions to UTF-8 that could be a large step, and one I would consider making separately, at a future time.  Any errors in this area could break key parts of wrapper.  I would suggest that it be done slowly, module by module.

**Comment by suresttexas00** [comment]
I agree that Python 3 handles text encoding much better.  However, wrapper was written heavily with Py2 ways of doing things.

**Comment by suresttexas00** [comment]
If you were just referring to adding `# -*- coding: utf-8 -*-` headers to the files, then that is great, I was thinking the same thing!

**Comment by JasonBristol** [comment]
Yah that is what I meant, I should have been clearer, but this along with
hashbangs will help standardize our code

**Comment by suresttexas00** [comment]
Another thing that is problematic is stuff like this:
```
 	def getClient(self):
		"""
		:returns: player client object
		"""
 		if self.client == None:
 			for client in self.wrapper.proxy.clients:
 				try:
					if client.username == self.username:
 						self.client = client
 						return self.client
 				except:
 					pass
 		else:
 			return self.client
```
In this case if self.client is None and no matching client is found, the method just returns nothing ("True"?).  In most places in the code (or where a plugin may be calling this method), the getClient statement is just "used" without any checking to see if a valid client is returned.  Since there is not "code" (or False) returned to signal a bad/missing client, there is really no way for plugins or wrapper code to test for a bad client except by using (very frequently) `try` - `except`'s

**Comment by JasonBristol** [comment]
Global variables seem to be a large crutch in the codeset, it is bad
practice since these can be manipulated by multiple parties, I think this
is where a lot of our issues originate. A good resource is The Hitchhiker's
Guide to Python <http://docs.python-guide.org/> which states:

   - Heavy usage of global state or context: instead of explicitly passing
   (height, width, type,wood) to each other, Table and Carpenter rely on
   global variables that can be modified and are modified on the fly by
   different agents. You need to scrutinize all access to these global
   variables to understand why a rectangular table became a square, and
   discover that remote template code is also modifying this context, messing
   with table dimensions.

There is a ton of spaghetti code as well, I think my favorite example is in
plugins.py:

try: disabled = plugin.DISABLED
except: disabled = False
try: dependencies = plugin.DEPENDENCIES  # if used, plugin.DEPENDENCIES
must be a 'list' type (even if only one item); e.g. = ["some.py",
"another.py", etc]
except: dependencies = False
try: name = plugin.NAME
except: pass
try: id = plugin.ID
except: id = name
try: version = plugin.VERSION
except: version = (0, 1)
try: description = plugin.DESCRIPTION
except: description = None
try: summary = plugin.SUMMARY
except: summary = None
try: author = plugin.AUTHOR
except: author = None
try: website = plugin.WEBSITE
except: website = None

**Comment by suresttexas00** [comment]
We should exercise some diplomacy here..


and edit our comments accordingly.. but here in an email we can be a little more open about it.


Here is a snippet of Ben's position on PEP-8 and Tabs in particular.  I went around with him about this before.  I even created a pep-8 pull request which was completely ignored.


this discussion comes from issue #190 on his fork:


"""

'll have to review it a bit more, but one thing that I'll admit that I'm 100% stubborn about is tabs vs spaces. I hate spaces so much for indentation, and the vast majority of arguments for "why you shouldn't use tabs" are easily reversible to "why you shouldn't use spaces".

e.g. "You shouldn't mix tabs and spaces" - okay, that doesn't mean I should use spaces instead, that just tells me that I shouldn't mix them.

"Tabs don't have a defined width" - the spaces method aren't either. you can have one space, two spaces, three spaces, four spaces, five spaces, etc. etc. etc.

My arguments against spaces:

  *   Pressing the space bar four times sucks. Some editors will convert tabs-to-spaces, but they aren't plentiful and my arguments against this are below.
  *   Making uber-quick edits in nano or another simplistic editor aren't as easy, for the aforementioned reason. I like to be able to open my code up in any editor that I have available (if I'm not at my workstation, for example) and do what I need with minimal hassle.
  *   I've never seen an editor that hasn't had the ability to adjust tab width, besides maybe crap like Microsoft Notepad. However, I've seen editors that lack the ability to convert tabs to spaces easily.
  *   PEP-8 is just a standard for Python's internal module code - there's no reason to follow it.
  *   ME IS STUBBORN

I've seriously never had an issue with tabs, and spaces is just more hassle, so I'd really rather not go that route. The other stuff, on the other hand, I'll have to evaluate.

"""


I would like to get him out of this mindset but saying things like your "professor would give you a 0 for .." will not help our cause.


I would like to make our case better by:

1) the fact that he does not have to make "uber quick edits i nano" anymore now.

2) the wrapper improvements speak for themselves.


just bear in mind "HE IS STUBBORN".


until he comes into the 21st century on tabs... well, I don't want to think about that.


Just remember to be gracious and diplomatic.  "We'll catch more flies with honey.."

**Comment by suresttexas00** [comment]
I think there has been enough progress here to justify closing this specific thread issue?  Additional or specific items can be opened up as new threads..?
